src/main/java/org/nuxeo/audit/storage/impl/SQLAuditStorage.java

```diff
@@ -0,0 +1,196 @@
+	}
+
+	/**
+	 * Scroll log entries in the SQL Audit directory, given an audit query
+	 * builder.
+	 */
+	@Override
+	public ScrollResult scroll(AuditQueryBuilder queryBuilder, int batchSize, int keepAlive) {
+		cursorService.checkForTimedOutScroll();
+		List<LogEntry> logEntries = queryLogs(queryBuilder);
+		String scrollId = cursorService.registerCursor(logEntries.iterator(), batchSize, keepAlive);
+		return scroll(scrollId);
+	}
+
+	/**
+	 * Query log entries in the SQL Audit directory, given an audit query
+	 * builder. Does not support literals other than StringLiteral: see
+	 * {@link SQLSession.query(Map<String, Serializable>, Set<String>,
+	 * Map<String, String>, boolean, int, int)};
+	 */
+	protected List<LogEntry> queryLogs(AuditQueryBuilder queryBuilder) {
+		List<LogEntry> logEntries = new ArrayList<LogEntry>();
+
+		// Get the predicates filter map from the query builder.
+		Map<String, Serializable> filter = new HashMap<String, Serializable>();
+		MultiExpression predicate = (MultiExpression) queryBuilder.predicate();
+		@SuppressWarnings("unchecked")
+		List<Predicate> predicateList = (List<Predicate>) ((List<?>) predicate.values);
+		for (Predicate p : predicateList) {
+			String rvalue;
+			if (p.rvalue instanceof StringLiteral) {
+				rvalue = ((StringLiteral) p.rvalue).asString();
+			} else {
+				rvalue = p.rvalue.toString();
+				log.warn(String.format(
+						"Scrolling audit logs with a query builder containing non-string literals is not supported: %s.",
+						rvalue));
+			}
+			filter.put(p.lvalue.toString(), rvalue);
+		}
+
+		// Get the orderBy map from the query builder.
+		Map<String, String> orderBy = new HashMap<String, String>();
+		ArrayList<OrderByExpr> orders = queryBuilder.orders();
+		orders.stream().forEach(order -> orderBy.put(order.reference.name, order.isDescending ? "DESC" : "ASC"));
+
+		// Get the limit and offset from the query builder.
+		int limit = (int) queryBuilder.limit();
+		int offset = (int) queryBuilder.offset();
+
+		// Query the Json Entries via the SQL directory session.
+		SQLSession session = (SQLSession) sqlDirectory.getSession();
+		DocumentModelList jsonEntriesDocs = session.query(filter, Collections.singleton(JSON_COLUMN), orderBy, false,
+				limit, offset);
+		session.close();
+
+		// Build a list of Log Entries from the Json Entries.
+		String auditPropertyName = sqlDirectory.getSchema() + ":" + JSON_COLUMN;
+		for (DocumentModel jsonEntryDoc : jsonEntriesDocs) {
+			String jsonEntry = String.valueOf(jsonEntryDoc.getPropertyValue(auditPropertyName));
+			LogEntry logEntry = getLogEntryFromJson(jsonEntry);
+			if (logEntry != null) {
+				logEntries.add(logEntry);
+			}
+		}
+
+		return logEntries;
+	}
+
+	/**
+	 * Convert a Json entry to a LogEntry.
+	 */
+	protected LogEntry getLogEntryFromJson(String jsonEntry) {
+		try {
+			ObjectMapper mapper = new ObjectMapper();
+			mapper.getDeserializationConfig().addMixInAnnotations(LogEntryImpl.class, LogEntryImplMixIn.class);
```

Comment:
Constructing an `ObjectMapper` for each entry is costly, you should do it only once. `ObjectMapper` is thread-safe so you can construct it statically:

    protected static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    static {
        OBJECT_MAPPER.getDeserializationConfig().addMixInAnnotations(LogEntryImpl.class, LogEntryImplMixIn.class);
    }